### PR TITLE
Грудзин Константин. Задача 3. Вариант 10. Интегрирование методом Монте-Карло многомерных интегралов

### PR DESCRIPTION
**Pull request by Konstantin-Grudzin**
Основная идея многопоточного решения:
Статистически, маловероятно попадание нескольких потоков в один и тот же вектор (Особенно при разных сидах)
Поэтому достаточно каждому потоку просуммировать отведённое количество случайных точек и слить их в нулевом потоке
Чтобы удостоверится, что теория верна, дополнительно написана последовательная версия, с которой сравнивается программа (на ошибку не более одного процента от обычного метода Монте-Карло)

### CODE REVIEW

File: tasks/mpi/grudzin_k_monte_carlo/func_tests/gmc_ftests.cpp [hunk 1]
```diff
@@ -0,0 +1,323 @@
+#include <gtest/gtest.h>
```
Comment (Napkin-AI) [inline tasks/mpi/grudzin_k_monte_carlo/func_tests/gmc_ftests.cpp:1]: Add sequential vesion to folder seq to measure performance

File: tasks/mpi/grudzin_k_monte_carlo/func_tests/gmc_ftests.cpp [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+#include <boost/mpi/communicator.hpp>
+#include <boost/mpi/environment.hpp>
+#include <vector>
+
+#include "mpi/grudzin_k_monte_carlo/include/gmc_include.hpp"
+
+namespace grudzin_k_montecarlo_mpi {
+std::vector<double> GenDimDistr(int dim) {
+  std::mt19937 rnd;
+  std::uniform_real_distribution<> dist(-5.0, 5.0);
```
Comment (Napkin-AI) [inline tasks/mpi/grudzin_k_monte_carlo/func_tests/gmc_ftests.cpp:12]: fix the seed

File: tasks/mpi/grudzin_k_monte_carlo/func_tests/gmc_ftests.cpp [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+    int finish = dist(rnd);
+    if (finish < start) std::swap(start, finish);
+    tmp.emplace_back(start);
+    tmp.emplace_back(finish);
+  }
+  return tmp;
+}
+}  // namespace grudzin_k_montecarlo_mpi
+
+TEST(grudzin_k_monte_carlo_mpi, Test_1Dim) {
```
Comment (Napkin-AI) [inline tasks/mpi/grudzin_k_monte_carlo/func_tests/gmc_ftests.cpp:25]: try N = 2^k,  (for expl k = 15)

File: tasks/mpi/grudzin_k_monte_carlo/func_tests/gmc_ftests.cpp [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+  auto f = [](std::array<double, 3> &x) -> double { return std::pow(x[0], 3) + std::pow(x[1] + x[2], 2) + 2.0 * x[2]; };
+  std::vector<double> dim = grudzin_k_montecarlo_mpi::GenDimDistr(dimensions);
+  double result_par = 0;
+  double result_seq = 0;
+
+  if (world.rank() == 0) {
+    std::shared_ptr<ppc::core::TaskData> MC1_Data = std::make_shared<ppc::core::TaskData>();
+    std::shared_ptr<ppc::core::TaskData> MC2_Data = std::make_shared<ppc::core::TaskData>();
+    MC1_Data->inputs.emplace_back(reinterpret_cast<uint8_t *>(dim.data()));
+    // MC1_Data->inputs.emplace_back(reinterpret_cast<uint8_t *>(&N));
```
Comment (Napkin-AI) [inline tasks/mpi/grudzin_k_monte_carlo/func_tests/gmc_ftests.cpp:273]: remove the comments please, your code is hard to read

File: tasks/mpi/grudzin_k_monte_carlo/include/gmc_include.hpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  bool pre_processing() override;
+  bool validation() override;
+  bool run() override;
+  bool post_processing() override;
+
+ private:
+  // range of every dimension
+  std::vector<double> dim;
+  std::function<functionData> f;
+  // answer
```
Comment (Napkin-AI) [inline tasks/mpi/grudzin_k_monte_carlo/include/gmc_include.hpp:34]: remove the comment. I see the variable with name "result" is an answer

File: tasks/mpi/grudzin_k_monte_carlo/include/gmc_include.hpp [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  bool validation() override;
+  bool run() override;
+  bool post_processing() override;
+
+ private:
+  // range of every dimension
+  std::vector<double> dim;
+  // target function
+  std::function<functionData> f;
+  // answer
```
Comment (Napkin-AI) [inline tasks/mpi/grudzin_k_monte_carlo/include/gmc_include.hpp:54]: remove the comment. I see the variable with name "result" is an answer

File: tasks/mpi/grudzin_k_monte_carlo/include/gmc_include.hpp [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+      : Task(std::move(taskData_)), f(f_) {}
+  bool pre_processing() override;
+  bool validation() override;
+  bool run() override;
+  bool post_processing() override;
+
+ private:
+  // range of every dimension
+  std::vector<double> dim;
+  // target function
```
Comment (Napkin-AI) [inline tasks/mpi/grudzin_k_monte_carlo/include/gmc_include.hpp:52]: either remove this comment or add it to seq version

File: tasks/mpi/grudzin_k_monte_carlo/src/gmc_src.cpp [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+template <const int dimension>
+bool MonteCarloSeq<dimension>::validation() {
+  internal_order_test();
+  return !taskData->outputs_count.empty() && taskData->outputs_count[0] == 1 && taskData->inputs_count.size() == 2 &&
+         taskData->inputs_count[0] == dimension;
+}
+
+template class MonteCarloMpi<1>;
```
Comment (Napkin-AI) [inline tasks/mpi/grudzin_k_monte_carlo/src/gmc_src.cpp:125]: what is that?

**Review by Napkin-AI — CHANGES_REQUESTED** [review]

File: tasks/mpi/grudzin_k_monte_carlo/func_tests/gmc_ftests.cpp (hunk 2 above)
Comment (Konstantin-Grudzin) [reply tasks/mpi/grudzin_k_monte_carlo/func_tests/gmc_ftests.cpp:12]: What do you want to fix? If you construct it by default (without specified seed) it use default seed (like zero or smth like that). I can't use some random seed cause' it bring unpredictable situations. (if you interested, i test it on different seeds and it worked)

File: tasks/mpi/grudzin_k_monte_carlo/func_tests/gmc_ftests.cpp (hunk 3 above)
Comment (Konstantin-Grudzin) [reply tasks/mpi/grudzin_k_monte_carlo/func_tests/gmc_ftests.cpp:25]: add power of two and prime number for count of dots

File: tasks/mpi/grudzin_k_monte_carlo/src/gmc_src.cpp (hunk 8 above)
Comment (Konstantin-Grudzin) [reply tasks/mpi/grudzin_k_monte_carlo/src/gmc_src.cpp:125]: Without it my code dont see this versions of function, i cnnot remove that

File: tasks/mpi/grudzin_k_monte_carlo/include/gmc_include.hpp [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+#include <boost/mpi/collectives.hpp>
+#include <boost/mpi/communicator.hpp>
+#include <boost/serialization/vector.hpp>
+#include <memory>
+#include <numeric>
+#include <random>
+#include <string>
+#include <utility>
+#include <vector>
```
Comment (m1likus) [inline tasks/mpi/grudzin_k_monte_carlo/include/gmc_include.hpp:13]: Some of this includes already are in task.hpp

File: tasks/seq/grudzin_k_monte_carlo/func_tests/main.cpp [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+  MC2_Data->inputs_count.emplace_back(1);
+  MC2_Data->outputs.emplace_back(reinterpret_cast<uint8_t *>(&result_seq));
+  MC2_Data->outputs_count.emplace_back(1);
+
+  grudzin_k_montecarlo_seq::MonteCarloSeq<dimensions> MC2(MC2_Data, f);
+  ASSERT_EQ(MC2.validation(), true);
+  MC2.pre_processing();
+  MC2.run();
+  MC2.post_processing();
+  EXPECT_LE(abs(etalon - result_seq) / etalon, 1e-1);
```
Comment (m1likus) [inline tasks/seq/grudzin_k_monte_carlo/func_tests/main.cpp:54]: Why do you use Less Equal everywhere and not Near?

**Review by m1likus — CHANGES_REQUESTED** [review]

**Review by Napkin-AI — APPROVED** [review]

File: tasks/seq/grudzin_k_monte_carlo/func_tests/main.cpp (hunk 10 above)
Comment (Konstantin-Grudzin) [reply tasks/seq/grudzin_k_monte_carlo/func_tests/main.cpp:54]: Near just checked,that your value located between some diapasone of dot ([a-x;a+x]), but absolute error can be less that one precent on some value, i can make smth like that: (my statement,0.1,0.1), but it's much more text

Comment (m1likus) [reply tasks/seq/grudzin_k_monte_carlo/func_tests/main.cpp:54]: I think at least at sec impl near is more logic than LE but that's fine then now I get it

**Review by m1likus — APPROVED** [review]

**Review by aobolensk — APPROVED** [review]

File: tasks/mpi/grudzin_k_monte_carlo/func_tests/gmc_ftests.cpp [hunk 11]
```diff
@@ -0,0 +1,411 @@
+#include <gtest/gtest.h>
+
+#include <boost/mpi/communicator.hpp>
+#include <boost/mpi/environment.hpp>
+#include <vector>
+
+#include "mpi/grudzin_k_monte_carlo/include/gmc_include.hpp"
+
+namespace grudzin_k_montecarlo_mpi {
+std::vector<double> GenDimDistr(int dim) {
```
Comment (allnes) [inline tasks/mpi/grudzin_k_monte_carlo/func_tests/gmc_ftests.cpp:10]: please parametrize range

File: tasks/mpi/grudzin_k_monte_carlo/func_tests/gmc_ftests.cpp [hunk 12]
```diff
@@ -0,0 +1,411 @@
...
+    EXPECT_LE(abs(result_par - result_seq) / result_seq, 1e-1);
+  }
+}
+
+TEST(grudzin_k_monte_carlo_mpi, Test_2Dim) {
+  boost::mpi::communicator world;
+  const int dimensions = 2;
+  int N = 10000;
+  std::function<double(std::array<double, dimensions> &)> f = [](std::array<double, dimensions> &x) {
+    { return 0.0 * (x[0] + x[1]) + 1.0; }
```
Comment (allnes) [inline tasks/mpi/grudzin_k_monte_carlo/func_tests/gmc_ftests.cpp:73]: please don't use functions with zero multiplication

Comment (Konstantin-Grudzin) [reply tasks/mpi/grudzin_k_monte_carlo/func_tests/gmc_ftests.cpp:73]: Sorry, but i cannot delete this: one of compilators complaning about unusing variables, this multiplication needs for accepting this issue

Comment (Konstantin-Grudzin) [reply tasks/mpi/grudzin_k_monte_carlo/func_tests/gmc_ftests.cpp:73]: (It was clang build)

**Pull request merged**